_posts/2017-10-25-tips-for-upgrading-rails-3-2-to-4.markdown

```diff
@@ -0,0 +1,212 @@
+| devise-encryptable 0.2.0 -> devise 2.1.4 | railties ~> 3.1            |
+| friendly_id 4.0.10.1                     | activerecord < 4.0, >= 3.0 |
+| strong_parameters 0.2.3                  | actionpack ~> 3.0          |
+| strong_parameters 0.2.3                  | activemodel ~> 3.0         |
+| strong_parameters 0.2.3                  | activesupport ~> 3.0       |
+| strong_parameters 0.2.3                  | railties ~> 3.0            |
++------------------------------------------+----------------------------+
+```
+
+Instead of going through your currently bundled gems or `Gemfile.lock` manually,
+you get a report of the gems you need to upgrade.
+
+<h2 id="config-files">3. Config files</h2>
+
+Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task).
+You can use this task as a guideline as explained thoroughly in
+[this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html).
+It will help you get rid of unnecessary code or monkey-patches in your config
+files and initializers, specially if your Rails 3 app was running on Rails 2.
+
+As an alternative, check out [RailsDiff](http://railsdiff.org/3.2.22.5/4.0.13),
+which provides an overview of the changes in a basic Rails app between 3.2 and
+4.0 (or any other source/target versions).
+
+If you're feeling adventurous, you
+can give [this script](https://github.com/bsodmike/upgrade_rails_3.2.12_to_4.0.0.beta1)
+a try. It attempts to apply [this git patch](https://github.com/bsodmike/upgrade_rails_3.2.12_to_4.0.0.beta1/blob/master/upgrade/upgrade.patch)
+(similar to the patch shown on RailsDiff) to your Rails app to migrate from 3.2
+to 4.0. However, I don't recommend this for complex or mature apps, as there
+will be plenty of conflicts.
+
+<h2 id="application-code">4. Application code</h2>
+
+<h2 id="models">a. Models</h2>
+
+- All dynamic finder methods except for `.find_by_...` are deprecated:
+
+```ruby
+# before:
+Authentication.find_all_by_provider_and_uid(provider, uid)
+
+# after:
+Authentication.where(provider: provider, uid: uid)
+```
+
+You can regain usage of these finders by adding the gem
+[activerecord-deprecated_finders](https://github.com/rails/activerecord-deprecated_finders)
+
+- ActiveRecord scopes now need a lambda:
+
+```ruby
+# before:
+default_scope where(deleted_at: nil)
+
+# after:
+default_scope { where(deleted_at: nil) }
+
+# before:
+has_many :posts, order: 'position'
+
+# after:
+has_many :posts, -> { order('position') }
+```
+
+(Friendly reminder: beware when using [default_scope](https://www.ombulabs.com/blog/ruby/rails/best-practices/why-using-default-scope-is-a-bad-idea.html))
+
+- Protected attributes is deprecated, but you can still add the [protected_attributes](https://github.com/rails/protected_attributes) gem.
+However, since the Rails core team dropped its support since Rails 5.0, you
+should begin migrating your models to Strong Parameters anyway.
```

Comment:
@mauro-oto please add a link to http://guides.rubyonrails.org/action_controller_overview.html#strong-parameters